경북대 FE 장세현 2단계 - 선물하기 메인 페이지 UI 구현하기

SeHIgh opened the pull request:
## STEP 2

안녕하십니까 경북대학교 프론트엔드 과정 수강중인 장세현입니다.
2단계 과제 - 선물하기 메인 페이지 UI 구현하기 내용 PR 요청 드립니다!

### 💫 과제 진행 요구사항 및 작업 내용

- [x] 본인만의 기준으로 일관된 코드 작성
  - [x] 과도하지 않게 적절히 컴포넌트 나누어 재사용하기
- [x] 기능 단위로 나누어 커밋   
       <img width="731" alt="image" src="https://github.com/user-attachments/assets/659e0eaa-5706-4463-b35b-ef3985db7cbd" />
- [x] emotion에 Color, Typography 토큰 설정
      (토큰 가이드 : https://edu.nextstep.camp/s/0eoNzeZS/ls/lNeUqOll)
- [x] Mobile First Design 기반의 Layout 구현
  - [x] 화면의 max-width를 720px으로 제한
- [x] 상단 네비게이션 바 영역 구현
- [x] 카테고리 섹션 구현
  - [x] Flex, Grid 등을 활용
- [x] 실시간 급상승 선물랭킹 구현
  - [x] Flex, Grid 등을 활용
- [x] 기타 섹션 (선물한 친구 선택, 카테캠 화이팅) UI 구현
- [x] 설정한 테마 토큰으로 리팩토링하기 (임의로 설정한 색상 변경)

### 전체 UI
![image](https://github.com/user-attachments/assets/5ebce332-3865-4481-8596-4c20dd9bbb19)
![image](https://github.com/user-attachments/assets/4860d786-79c1-4ce8-9472-abee7ddd7465)

### 추가 전달 사항
- 이전 1단계 병합 이후 step 2 브랜치 생성과정에서 충돌 문제가 있어
   rebase 작업 후 기능구현을 진행했는데, 이후 PR 시 이전 커밋들도 같이 포함이 되는 것 같습니다.
   이부분에 대해서 어떻게 해결하면 좋을지 궁금합니다.
![image](https://github.com/user-attachments/assets/a8e4f993-0d1d-4a26-8a01-197c7a4866eb)

---
- 죄송합니다! 북치고 장구치는 것 같지만, 혼자서 해결해보았습니다!
- upstream 의 브랜치 로부터 새로운 commit 정리작업용 브랜치(step2-fix)를 origin 에 생성하였고, 
- cherry-pick 을 이용하여 2단계 작업 내용들을 적절히 그대로 다 가져온뒤 commit을 깔끔하게 정리하였습니다.
- 그리고 force push 를 통해 step2 브랜치 내용을 덮어써서
- 충돌 문제 그리고 중복 커밋 메시지가 남지 않도록 해결하였습니다!

SeHIgh committed:
docs(README): 2단계 과제 진행 요구사항 추가 및 구체화

SeHIgh committed:
feat(router): 추후 페이지 라우팅을 위한 Router 추가

- react-router-dom 라이브러리 설치
- src/App.tsx 수정
- src/pages/Home.tsx 추가
- src/shared/Router.tsx 추가

SeHIgh committed:
feat(theme): emotion 테마 적용 및 Home 페이지 스타일 개선

- README.md에 emotion Color, Typography 토큰 설정 완료 표시
- tsconfig 파일에 jsxImportSource 추가 (tsconfig.app.json 추가)
- main.tsx에서 ThemeProvider로 emotion 테마 전역적으로 적용
- Home.tsx에서 useTheme 훅을 사용하여 스타일 적용

SeHIgh committed:
fix(theme, friends, banner): 충돌 해결

- README.md에 emotion Color, Typography 토큰 설정 완료 표시
- tsconfig 파일에 jsxImportSource 추가 (tsconfig.app.json 추가)
- main.tsx에서 ThemeProvider로 emotion 테마 전역적으로 적용
- Home.tsx에서 useTheme 훅을 사용하여 스타일 적용
- 기타 배너 및 선물할 친구 섹션 추가

SeHIgh committed:
feat(Home): Mobile First Design 기반의 Layout 구현 및 네비게이션 바 영역 구현

- package.json 및 package-lock.json에 material-icons 의존성 추가
- Home.tsx에서 styled-components를 사용하여 페이지 구조 개선
- base.ts 추가하여 baseStyle 도 추가적으로 전역적으로 설정할 수 있도록 구현
- reset.ts에 링크 및 버튼 기본 스타일 초기화 추가
- README.md에 카카오톡 선물하기 참고 페이지 링크 추가
- 네비게이션 바 영역 구현
  - 뒤로가기 버튼, 프로필 버튼, 및 홈 이동 타이틀 추가

SeHIgh committed:
feat(Category): 카테고리 섹션 구현

- 카테고리 섹션 컴포넌트 구현
- Flex 및 Grid 사용하여 레이아웃 구현
- 추후 확장을 위해 카테고리 각 아이템을 Link로 구현
- MOCK 데이터 추가 (카테고리 섹션, 상품 목록)

SeHIgh committed:
feat(Trend): 실시간 급상승 섹션 추가 및 네비게이션 바 고정

- 네비게이션 바를 화면 상단에 고정하도록 스타일 수정
- 홈 페이지에 실시간 급상승 섹션 추가
- 상품 목록에 MOCK 데이터 추가 및 기본 스타일 설정 개선
- README.md에 설정한 테마 토큰으로 리팩토링 항목 추가

SeHIgh committed:
feat(friends, banner): 기타 배너 및 선물할 친구 섹션 추가

kyoungsik-dev commented on src/components/home/Trend/ProductItem.tsx:
```diff
@@ -0,0 +1,141 @@
+  product,
+  index,
+}: {
+  product: (typeof productListMock)[number]
```
지금은 Mock 데이터를 가지고 있으니 이처럼 typeof를 이용해서 타입을 지정할 수 있는데요,
실제 데이터를 붙이게 되면 이와 같은 방식은 불가능합니다.
typescript 컴파일하는 시점에는 api의 응답값을 알 수 없기 때문이지요.
때문에 상품정보를 담을 수 있는 별도의 type 또는 interface를 생성해주시는게 좋겠습니다 😄

kyoungsik-dev commented on src/shared/Router.tsx:
shared 디렉토리 아래에 Router 파일을 두었는데요,
shared 라는 이름은 공유자원, 예를 들면 공통 컴포넌트나 유틸리티 함수, 공통 상수 등 여러 모듈에서 공통적으로 사용되는 코드를 두기 위한 디렉토리에 적합해보입니다.
반면 Router 같은 경우는 App.tsx 파일 한 곳에서만 사용되는 컴포넌트이기 때문에 적합한 네이밍이 아니어 보입니다.

kyoungsik-dev commented on src/shared/Router.tsx:
```diff
@@ -0,0 +1,14 @@
+    <BrowserRouter>
+      <Routes>
+        <Route path="/" element={<Home />} />
+      </Routes>
```
지금은 Router 컴포넌트 안에 라우팅 정보가 모두 하드코딩되어있는데요,
만약 규모가 커지게 되면 이러한 방식으로는 사용하기가 어려울 수 있습니다.

```
// routes.ts
import { Home } from '@/pages/Home'

export const routes = [
  { path: '/', element: <Home /> },
]
```
따라서 라우팅 정보는 위와 같이 별도 상수로 분리하고,

```suggestion
    <Routes>
      {routes.map(({ path, element }) => (
        <Route key={path} path={path} element={element} />
      ))}
    </Routes>
```
위와 같이 해당 데이터를 활용하여 react-router-dom에 적용하는 것도 좋은 방법일 것 같습니다.

다만 현재는 경로가 하나 뿐이라 ㅎㅎ
지금 이와 같은 방식을 적용할 필요는 없구요, 그냥 참고만 해주셔도 될 것 같습니다 :)

kyoungsik-dev commented:
<img width="455" alt="image" src="https://github.com/user-attachments/assets/a725839c-d3a6-4c07-b970-b4cc65f9d4ca" />

모바일 (375px) 화면에서는 이처럼 상품 그리드가 다소 깨져 보이는데요, 이 부분도 개선되면 좋을 것 같습니다!

kyoungsik-dev commented on src/components/home/Trend/Trend.tsx:
```diff
@@ -0,0 +1,169 @@
+
+// * 실시간 급상승 컴포넌트
+export const Trend = () => {
+  const [searchParams, setSearchParams] = useSearchParams()
```
searchParams 사용해서 구현하신 것 너무 멋지네요!
덕분에 필터 선택에 따라 경로가 유지될 수 있게 되었네요 👍👍👍

kyoungsik-dev commented on src/components/home/Trend/types.ts:
```diff
@@ -0,0 +1,13 @@
+// * 타입 정의
+export type TargetType = 'ALL' | 'FEMALE' | 'MALE' | 'TEEN'
+export type RankType = 'MANY_WISH' | 'MANY_RECEIVE' | 'MANY_WISH_RECEIVE'
```
너무 좋습니다!!👍
단순히 string으로 처리할 수도 있었을테지만요,
고정되는 값이다보니 이런식으로 타입을 미리 정해주시면 훨씬 안전하겠죠!!

kyoungsik-dev commented on src/components/home/Trend/Trend.tsx:
```diff
@@ -0,0 +1,169 @@
+  // URL 파라미터에서 초기값 가져오기
+  const getInitialTargetType = (): TargetType => {
+    const urlTargetType = searchParams.get('targetType')
+    return urlTargetType && isValidTargetType(urlTargetType) ? urlTargetType : 'ALL'
```
`isValidTargetType`
ㄴ 예외처리까지 해주시고 아주 좋습니다~

kyoungsik-dev commented on src/components/home/Trend/Trend.tsx:
```diff
@@ -0,0 +1,169 @@
+    const urlTargetType = searchParams.get('targetType')
+    const urlRankType = searchParams.get('rankType')
+
+    if (urlTargetType && isValidTargetType(urlTargetType) && urlTargetType !== targetType) {
```
```suggestion
    if (urlTargetType && isValidTargetType(urlTargetType)) {
```

` urlTargetType !== targetType`
사실 이 비교는 불필요합니다.
setState에서 같은 값이면 어차피 렌더링을 일으키지 않고, targetType 값은 string 타입, 즉 primitive 타입이기 때문에 값을 비교하는 것도 안전하고요.

SeHIgh replied:
뭔가 더 타입을 상세히 검증해보고 싶었는데, 불필요한 코드는 없애는게 좋을 것 같다는 것을 다시 느끼게 되었습니다. 조언 감사합니다!

kyoungsik-dev commented on src/components/home/Trend/types.ts:
```diff
@@ -0,0 +1,13 @@
+// * 유효한 RankType인지 확인하는 함수
+export const isValidRankType = (value: string): value is RankType => {
+  return ['MANY_WISH', 'MANY_RECEIVE', 'MANY_WISH_RECEIVE'].includes(value)
+}
```
다만 좀더 제안을 드리자면...

타입 string 값들이 중복되는게 조금 아쉬워서요 ㅎㅎ

```suggestion
// * 타입 정의
export const TARGET_TYPE_LIST = ['ALL', 'FEMALE', 'MALE', 'TEEN'] as const
export const RANK_TYPE_LIST = ['MANY_WISH', 'MANY_RECEIVE', 'MANY_WISH_RECEIVE'] as const
export type TargetType = (typeof TARGET_TYPE_LIST)[number]
export type RankType = (typeof RANK_TYPE_LIST)[number]

// * 유효한 TargetType인지 확인하는 함수
export const isValidTargetType = (value: string): value is TargetType => {
  return TARGET_TYPE_LIST.includes(value as TargetType)
}

// * 유효한 RankType인지 확인하는 함수
export const isValidRankType = (value: string): value is RankType => {
  return RANK_TYPE_LIST.includes(value as RankType)
}
```

as const 키워드를 이용해 이렇게 작성해주시면 중복을 최소화할 수 있습니다 😁

SeHIgh replied:
너무나 유익한 피드백 감사합니다!

kyoungsik-dev commented on src/data/productListMock.ts:
```javascript
export const productListMock = Array.from({ length: 20 }, (_, index) => ({
  id: index,
  name: 'BBQ 양념치킨+크림치즈볼+콜라1.25L',
  imageURL:
    'https://st.kakaocdn.net/product/gift/product/20231030175450_53e90ee9708f45ffa45b3f7b4bc01c7c.jpg',
  price: {
    basicPrice: 29000,
    discountRate: 10,
    sellingPrice: 26100,
  },
  brandInfo: {
    id: 2088,
    name: 'BBQ',
    imageURL:
      'https://st.kakaocdn.net/product/gift/gift_brand/20220216170226_38ba26d8eedf450683200d6730757204.png',
  },
}));
```

mock 데이터도 이렇게 반복문을 이용해서 만들어주면 좀더 nice 해보이지 않을까요?
사실 지금과 같이 해주셔도 전혀 상관은 없는데요,
id값이 중복되는게 조금 아쉬워서요.
render 함수에서는 product.id 값을 key로 사용하고 있는데요, key 값이 중복되다 보니 아래와 같이 에러가 발생하고 있습니다.
(개발 과정중에 이런식으로 뜨는 warning, error는 무시하지 말고 꼭 확인하는 습관도 필요합니다!)
<img width="575" alt="image" src="https://github.com/user-attachments/assets/62cdd8f3-c7c9-47f0-bae3-ac2eb3d50306" />
물론 실제 데이터를 연동하게 되면 이와 같은 에러는 발생하지 않겠지만,
개발과정 중이라도 이러한 에러가 쌓이게 되면, 다른 에러를 놓치게 될 수도 있으니까요 🥹

SeHIgh replied:
넵! 다음부터는 사소한 에러를 놓치지않도록 주의하겠습니다!

kyoungsik-dev commented on src/components/home/Banner.tsx:
```diff
@@ -0,0 +1,70 @@
+    <Container>
+      <SubContainer>
+        <SubBannerTitle>카카오테크 캠퍼스 3기여러분</SubBannerTitle>
+        <BannerTitle>프론트엔드 2단계 과제 화이팅! 🎉</BannerTitle>
```
배너 내 텍스트는 바뀌는 경우가 많겠죠?ㅎㅎ
하드코딩하기 보단 다른 방법을 고민해봐주세요.

kyoungsik-dev reviewed (changes requested):
안녕하세요!.
코드 퀄리티가 정말 높아서.. 사실 많이 놀랬습니다!
고민한 흔적이 많이 보였습니다.👍👍

코멘트를 많이 드리고 싶었는데 흠잡을 데가 없어서 쉽지 않았네요.ㅎㅎ
코멘트 남긴 내용 천천히 살펴보시고, 수정 부탁드리겠습니다.
정말 고생 많으셨고, 계속 응원하겠습니다!

SeHIgh committed:
fix(Trend): 모바일 (375px) 화면 상품 그리드 깨지는 현상 수정

- 상품 가격 컨테이너에 wrap 속성 추가

SeHIgh committed:
feat(Trend): 상품 인터페이스 및 타입 정의 추가

- MOCK 데이터 사용 안할 때를 가정하여 안전하게 인터페이스 추가
  - 상품 가격 및 브랜드 정보를 포함하는 Product 인터페이스 추가

SeHIgh committed:
docs(router): Router.tsx 속한 디렉토리명 변경

- src/shared -> src/router

SeHIgh committed:
fix(Trend): URL 파라미터 검증 로직 수정

- targetType 검증 로직에서 불필요한 조건 제거하여 코드 간소화

SeHIgh committed:
refactor(Trend): 타입 정의 및 검증 로직 개선

- TargetType 및 RankType을 상수 배열로 정의하여 코드 가독성 향상
- as const 키워드를 이용하여 string 값 중복 최소화

SeHIgh committed:
docs(MOCK): 상품 목록 MOCK 데이터 구조 개선

- 기존의 하드코딩된 상품 데이터를 배열 생성 방식으로 변경하여 코드 간소화
- 상품 ID를 동적으로 생성하여 중복을 방지

SeHIgh committed:
feat(Banner): 배너 제목을 하드코딩에서 데이터로 변경

- 배너 제목 및 서브 제목을 하드코딩에서 BANNER_CONTENT 데이터로 변경하여 유지보수성 향상

SeHIgh commented:
상세한 피드백 감사합니다. 피드백에 맞게 코드들을 수정해보았습니다!
1. 모바일 (375px) 화면 상품 그리드 깨지는 현상 수정
- `flex-wrap` 속성을 추가하여 width가 줄어들때 그에 맞게 깨지지 않도록 자동으로 배치되도록 하였습니다.
<img width="504" alt="image" src="https://github.com/user-attachments/assets/aee90747-8f13-4afe-9496-8537765cc086" />

2. 상품 인터페이스 및 타입 정의 추가
3. Router.tsx 속한 디렉토리명 변경
4. URL 파라미터 불필요한 검증 로직 수정
5. 타입 정의 및 검증 로직 개선 - `as const`키워드 활용
6. 상품 목록 MOCK 데이터 구조 개선
7. 배너 제목을 하드코딩에서 유지보수 가능한 데이터 형태로 변경

kyoungsik-dev reviewed (approved):
수정해주신 것 확인했습니다.
고생 많으셨습니다 😁😁😁
